instrumentation/kotlinx-coroutines/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kotlinxcoroutines/instrumentationannotations/WithSpanInstrumentation.java

```diff
@@ -0,0 +1,718 @@
+                  temp.visitVarInsn(parameter.type.getOpcode(Opcodes.ILOAD), parameter.var);
+                  boolean primitive =
+                      parameter.type.getSort() != Type.ARRAY
+                          && parameter.type.getSort() != Type.OBJECT;
+                  temp.visitMethodInsn(
+                      Opcodes.INVOKESTATIC,
+                      Type.getInternalName(AnnotationInstrumentationHelper.class),
+                      "setSpanAttribute",
+                      "(ILjava/lang/String;"
+                          + (primitive ? parameter.type.getDescriptor() : "Ljava/lang/Object;")
+                          + ")V",
+                      false);
+                }
+                // pop label
+                temp.visitInsn(Opcodes.POP);
+                if (hasBlockingOperation) {
+                  methodNode.instructions.insert(insertAfterInsn, temp.instructions);
+                } else {
+                  methodNode.instructions.insertBefore(
+                      methodNode.instructions.get(0), temp.instructions);
+                }
+              }
+
+              // insert at the start of the method
+              // null the local variables we added
+              // this is needed because jvm requires that a value needs to be assigned to the local
+              // before it is used, we need to initialize the locals that we use in the exception
+              // handler
+              // if the previous block was added at the start of the method this nulling step isn't
+              // necessary
+              if (hasBlockingOperation) {
+                MethodNode temp = new MethodNode();
+                temp.visitInsn(Opcodes.ACONST_NULL);
+                temp.visitVarInsn(Opcodes.ASTORE, requestLocal);
+                temp.visitInsn(Opcodes.ACONST_NULL);
+                temp.visitVarInsn(Opcodes.ASTORE, ourContinuationLocal);
+                temp.visitInsn(Opcodes.ACONST_NULL);
+                temp.visitVarInsn(Opcodes.ASTORE, contextLocal);
+                temp.visitInsn(Opcodes.ACONST_NULL);
+                temp.visitVarInsn(Opcodes.ASTORE, scopeLocal);
+
+                methodNode.instructions.insertBefore(
+                    methodNode.instructions.get(0), temp.instructions);
+              }
+
+              // insert exception handler code, this exception handler will run catch Throwable
+              {
+                MethodNode temp = new MethodNode();
+                // scopeLocal was the last local we added before the start of try block
+                int numLocals = scopeLocal + 1;
+                Object[] locals = new Object[numLocals];
+                // in this handler we are using only the locals we added, we don't care about method
+                // arguments and this, so we don't list them in the stack frame
+                for (int i = 0; i < locals.length - 2; i++) {
+                  locals[i] = Opcodes.TOP;
+                }
```

Comment:
Nit: you could zero out the entire array to avoid the magic number confusion; the last 4 entries are overwritten anyway
```suggestion
                Arrays.fill(locals, Opcodes.TOP);
```

Reply:
makes sense, I think I initially had less locals than in the final version so here it was `2` instead of `4`. An alternative would be to use aliases `firstLocal` and `lastLocal` to keep track where our locals are.